wip: draft porting guide (#81)
bao-project/bao-docs: source/bao_hyp/user_manual/porting.rst

Porting

> =======

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

> The Bao Hypervisor only contains architecture-specific drivers (e.g., interrupt controllers or
> IOMMUS). Therefore porting the hypervisor essentially resumes to two simple operation: creating a
> platform description folder containing a platform definition source and header files which
> indicates, among others, the number of CPUs or the memory available. The exception is an UART driver
> used fore logging, so if a driver is not available for a UART on the target platform this might

Reviewer 2: In my opinion, the first sentence doesn't contribute significantly. It looks like a bit disconnected from the remaining text.
```suggestion
Porting the hypervisor involves two basic steps: (i) creating a platform description folder with
source and header files that specify platform details, such as the number of CPUs or available
memory, and (ii) integrating a UART driver used for logging, so if a driver is not available for a
UART on the target platform this might require writing a custom driver. 
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

require writing such a driver. Finally, given for some platforms provide a custom mechanism to

> define DMA device Master IDs required for configuring IOMMUs, one might require to configure

Reviewer 2: At this point, it would be interesting to refer to some examples to contextualize the firmware changes

Reviewer: This will done in later sections.

Reviewer: I think that should be done in a later section dedicated to describing how to define DMA device master IDs

firmware to set these IDs. In sum, porting Bao to a platform requires:

1) Create a platform directory containing a platform description [mandatory];
2) Writing an UART driver [if the driver for the target UART is not available];

> 3) Setting up firmware to define DMA device Master IDs used by the platform IOMMU [platform dependent];

Reviewer 2: Here it would be useful to have a link to each section that describes the process.

> 3) Setting up firmware to define DMA device Master IDs used by the platform IOMMU [platform dependent];

Reviewer 3: I agree

> 3) Setting up firmware to define DMA device Master IDs used by the platform IOMMU [platform dependent];

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

Platform Description
--------------------

> Creating a new platform description for the Bao Hypervisor starts by creating a new directory under
> *src/platform*. For example, for port to a *foo* platform, create the directory *src/platform/foo*.

Reviewer 2: ```suggestion
Creating a new platform description for the Bao Hypervisor begins by establishing a new directory
under src/platform. For instance, when porting to a foo platform, generate the directory
src/platform/foo.
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

This directory includes ??? four files:

1) A source file (e.g., *src/platform/foo/foo_desc.c*) containing the platform description structure, `struct platform platform`
defininition;
2) A *plat.h* header file under *src/platform/foo/inc/plat/*  with platform-specific includes and definitions;
3) A *platform.mk* makefile containing definitions required by the build system (e.g., the target architecture);
4) A *objects.mk* containing the platform's target object files. In principle this should only contain the 

> object for the source description file, but you may add others if you have any other platform specific sources.

Reviewer 3: List is not working

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

> object for the source description file, but you may add others if you have any other platform specific sources.

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

Platform Description Structure
******************************

> A global structure of type `struct platform` named `platform` must be defined when porting Bao to a new platform.

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

CPU Description
###############

Besides the number of CPUs (**cpu_num**), a platform port might optionally define a fixed CPU master. For many platforms, at first only
one CPU starts executing and later this CPU is reponsible for waking up others through arch- and/or platform-specific means. In this case,
there is no concurrency issues during initialization. However, for platforms where all CPUs start executing simultaneously, concurrency 
issues might arise (e.g., this happens for Arm processors because atomic instructions cannot be used before MMU is enable). Here we require
the port to specificy which CPU is the CPU master, responsible for the initializaiton procedure, by setting **cpu_master_fixed** and **cpu_master**.

Memory Regions
##############

Platforms might provide a number of non-contigous memory regions. The port must describe the available memory regions by filling in the **region_num**
and **regions** fields. The former is the number of available regions while the latter defines an array of `struct mem_region` with the following fields:

.. code-block:: c

    struct mem_region {
        paddr_t base;
        size_t size;
        ...
    };

Where:

> - **base** [mandatory] - Is the base physical address of that memory region. Must be aligned to the minimum architecture's page size;
> - **size** [mandatory] - The size of the memory region. Must be aligned to the minimum architecture's page size;

Reviewer 2: ```suggestion
- **base** [mandatory] - Is the base physical address of that memory region.
- **size** [mandatory] - The size of the memory region.

.. warning::
    Ensure that the memory ``base`` and ``size`` are aligned to the minimum architecture's page size.
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

> Note that, by convention, we do not add on-chip/scratcpad SRAMs as a platform memory region as the hypervisor as an uniform view of memory.
> This regions should be assigned to guests in the form of devices.

Reviewer 2: ```suggestion
.. note::
  By convention, we do not add on-chip/scratcpad SRAMs as a platform memory region as the hypervisor as an uniform view of memory. This regions should be assigned to guests in the form of devices.
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

Cache Description
#################

TODO

> Architcture-Specific Platform Description
> #########################################

Reviewer: ```suggestion
Architecture-Specific Platform Description
##########################################
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

The architecture-specific field of the platform description includes fields that describe interrupt controllers or others. 

> .. tabs::

Reviewer 3: Tabs not working

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

                paddr_t gicc_addr;
                paddr_t gich_addr;
                paddr_t gicv_addr;
                paddr_t gicd_addr;
                paddr_t gicr_addr;

                irqid_t maintenance_id;
            } gic;

            struct smmu_dscrp {
                paddr_t base;
                streamid_t global_mask;
            } smmu;

            struct clusters {
                size_t num;
                size_t* core_num;
            } clusters;
        };

    Where, for the GIC interrupt controller `struct gic_dscrp` description:

    - **gic.gicc_addr** [mandatory for GICv2 platforms] - base address for the GIC's CPU Interface;
    - **gic.gich_addr** [mandatory for GICv2 platforms] - base address for the GIC's Virtual Interface Control Registers;
    - **gic.gicv_addr** [mandatory for GICv2 platforms] - base address for the GIC's Virtual CPU Interface;
    - **gic.gicd_addr** [mandatory] - base address for the GIC's Distributor;
    - **gic.gicr_addr** [mandatory for GICv3/4 platforms] - base address for the GIC's Redistributor;
    - **gic.maintenance_id** [mandatory] - The interrupt ID for the GIC's maintenance interrupt;

> 

Reviewer: I would split this for each GIC version to be more explicit.

```
For all platforms:
- **gic.maintenance_id**
- **gic.gicd_addr**

For GICv2 platforms:
- **gic.gicc_addr**
- **gic.gich_addr**
- **gic.gicv_addr**

For GICv3/4 platforms:
- **gic.gicr_addr**
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

    For the SMMU `struct smmu_dscrp`:

    - **smmu.base** [mandatory] - is the base address for the SMMU;
    - **smmu.global_mask** [optional; only valid for SMMUv2] - a mask to be applied to all SMMUv2's Stream Match Registers;

> 

Reviewer: Use inline code format for all code related text.

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

    Finally, when CPUs are organized in clusters, in the Arm architecture their IDs are assigned using an hierarchical schema.
    To be able to calculate the linearized ID for each core, we require the port to provide the number of CPUs of cluster in
    ascending order of AFF1.

> 

Reviewer: This needs an extra indentation to be grouped on each tab. Same for the RISC-V information.

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

                paddr_t base;      // Base address of the IOMMU mmapped IF
                irqid_t fq_irq_id; 
            } iommu;

            struct {
                paddr_t base;
            } aclint_sswi;
        };

    In case the available interrupt controller is the legacy PLIC:

    - **irqc.plic.base** [mandatory if PLIC is available] - is the base address for the PLIC;

    In case the available interrupt controller is an AIA contaning an APLIC:

    - **irqc.aia.aplic.base** [mandatory if APLIC is available] - is the base address for the APLIC;

    When an IOMMU is available:

    - **iommu.base** [mandatory if IOMMU is available] - is the base address for the IOMMU;
    - **iommu.fq_irq_id** [mandatory if IOMMU is available] - the Fault Queue interrupt ID (the current implementatio assumes this is a wired interrupt);

Platform Header
****************

The platform header contains any includes and definitions required for the target platform.
At a minimum it includes the header for the target UART driver.

For the RISC-V architecture it should define the following macros:

- `CPU_EXT_SSTC` if the target platform CPUs implement the SSTC extension.

> Platform Make Defintions
> ************************

Reviewer: ```suggestion
Platform Make Definitions
*************************
```

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.

The *platform.mk* file defines the following make variables:

- `ARCH` [mandatory] - Indicates the target architecture. Options: `armv8` or `riscv`;
- `platform_description` [mandatory] - the C source file containing the platform description (in our example *foo_desc.c*);
- `drivers` [mandatory] - a space-separated list of the drivers to be included, following the names of the *src/drivers/* directory. This usually just the target UART driver.
- `platform-cppflags` [optional] - any platform-specific flags to be passed to the pre-processor;
- `platform-cflags` [optional] - any platform-specific flags to be passed to the C compiler;
- `platform-asflags` [optional] - any platform-specific flags to be passed to the assembler;
- `platform-ldflags` [optional] - any platform-specific flags to be passed to the linker;

Depending on the architecture other variables must also be defined:

.. tabs::
    .. tab:: Arm

    - `GIC_VERSION` [mandatory] - indicates the GIC's version present on the platform. Option: `GICV2`, or `GICV3`.

    .. tab:: RISC-V

    - `IRQC` [mandatory] - indicates the interrupt controller available on the platform. Options: `PLIC`, `APLIC`, or `AIA`.

> 

Reviewer 3: Currently not working

Reviewer: Addressed in 04edb5e2a7e817ebaa2315ff49e106fe4a6360b4.
